cli/tools/upgrade.rs

```diff
@@ -15,12 +16,84 @@ use std::io::Write;
 use std::path::Path;
 use std::path::PathBuf;
 use std::process::Command;
+use std::time::Duration;
 
 static ARCHIVE_NAME: Lazy<String> =
   Lazy::new(|| format!("deno-{}.zip", env!("TARGET")));
 
 const RELEASE_URL: &str = "https://github.com/denoland/deno/releases";
 
+const UPGRADE_CHECK_INTERVAL: i64 = 12;
+
+const UPGRADE_CHECK_FETCH_DELAY: Duration = Duration::from_millis(500);
+
+pub fn check_for_upgrades(cache_dir: PathBuf) -> Option<String> {
+  if env::var("DENO_NO_UPDATE_CHECK").is_ok() {
+    return None;
+  }
+
+  let p = cache_dir.join("latest.txt");
+  let content = match std::fs::read_to_string(&p) {
+    Ok(file) => file,
+    Err(_) => "".to_string(),
+  };
+
+  let (mut last_checked, mut latest_version) = (None, None);
+  let split_content = content.split('!').collect::<Vec<_>>();
+  if split_content.len() == 2 {
+    last_checked = Some(split_content[0]);
+    latest_version = Some(split_content[1].to_owned());
+  }
```

Comment:
parsing and serialization of this file's contents should be separate functions with a unit test.